In [22]:
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.ensemble import RandomForestRegressor

df = pd.read_csv('hf://datasets/nprak26/remote-worker-productivity/remote_work_productivity.csv')
df

Employee_ID  Age  Years_Experience  WFH_Days_Per_Week      Gender  \
0        EMP0001   39                10                  2      Female   
1        EMP0002   33                 4                  5      Female   
2        EMP0003   40                 3                  3        Male   
3        EMP0004   48                14                  3        Male   
4        EMP0005   32                 6                  5        Male   
...          ...  ...               ...                ...         ...   
1495     EMP1496   53                 5                  5        Male   
1496     EMP1497   53                 9                  3  Non-binary   
1497     EMP1498   45                 2                  4        Male   
1498     EMP1499   44                12                  4      Female   
1499     EMP1500   40                 4                  5        Male   

       Education_Level Marital_Status Has_Children Location_Type  \
0     Associate Degree        Married          Yes         Urban   
1        Master Degree        Married           No         Urban   
2                  PhD         Single          Yes         Rural   
3      Bachelor Degree        Married          Yes         Urban   
4          High School       Divorced          Yes         Rural   
...                ...            ...          ...           ...   
1495     Master Degree        Married           No         Urban   
1496     Master Degree        Married          Yes      Suburban   
1497  Associate Degree         Single          Yes      Suburban   
1498   Bachelor Degree         Single          Yes      Suburban   
1499   Bachelor Degree       Divorced          Yes      Suburban   

            Department  ... Quality_Score Innovation_Score Efficiency_Rating  \
0              Product  ...          58.1             52.1              72.1   
1     Customer Success  ...          93.3             77.9              89.5   
2           Operations  ...          84.7             63.2              95.0   
3              Finance  ...          67.8             82.5              95.0   
4          Engineering  ...          86.4             67.5              95.0   
...                ...  ...           ...              ...               ...   
1495                HR  ...          84.1             58.3              89.4   
1496           Finance  ...          74.7             54.2              87.2   
1497       Engineering  ...          88.4             71.2              95.0   
1498        Operations  ...          93.9             88.6              95.0   
1499  Customer Success  ...          64.8             64.9              81.0   

     Meetings_Per_Week Commute_Time_Minutes  Job_Satisfaction Stress_Level  \
0                    4                   48              55.9            6   
1                   12                    0              96.1            3   
2                   15                   24              90.4            5   
3                    8                    8             100.0           10   
4                   10                    0             100.0            3   
...                ...                  ...               ...          ...   
1495                 9                    0              71.4            8   
1496                 3                   11              95.8            8   
1497                11                   34             100.0            4   
1498                10                    6             100.0            3   
1499                12                    0             100.0            5   

     Work_Life_Balance  Survey_Date  Response_Quality  
0                    8   2024-04-05            Medium  
1                    8   2024-01-29              High  
2                    6   2024-01-18              High  
3                    5   2024-04-18              High  
4                    4   2024-02-19              High  
...                ...          ...               ...  
1495     

In [23]:
# Remove label column and unecessary features
X = df.drop(['Internet_Speed_Category','Response_Quality', 'Productivity_Score', 'Employee_ID'], axis=1)
y = df['Productivity_Score']

# 80/20 split
X_train, X_test, y_train, y_test = train_test_split(
    X, y,
    test_size=0.2,
    random_state=42
)

In [24]:
from sklearn.pipeline import Pipeline
from sklearn.impute import SimpleImputer
from sklearn.preprocessing import StandardScaler, OneHotEncoder
from sklearn.compose import ColumnTransformer

# Separate categorical and numerical features
cat_features = X_train.select_dtypes(include='object').columns.tolist()
num_features = X_train.select_dtypes(exclude='object').columns.tolist()

# Numeric pipeline for scaler
num_pipeline = Pipeline(steps=[
    ('imputer', SimpleImputer(strategy='median')),
    ('scaler', StandardScaler())
])

# Categorical pipeline for onehot encoding
cat_pipeline = Pipeline(steps=[
    ('imputer', SimpleImputer(strategy='most_frequent')),
    ('onehot', OneHotEncoder(handle_unknown='ignore'))
])

# Combine pipelines
preprocessor = ColumnTransformer(transformers=[
    ('num', num_pipeline, num_features),
    ('cat', cat_pipeline, cat_features)
])

# Add regressor model
reg = Pipeline(steps=[
    ('preprocessor', preprocessor),
    ('regressor', RandomForestRegressor(random_state=42))
])

In [25]:
from sklearn.model_selection import GridSearchCV

# hyperparameter grid for Random Forest Regressor
param_grid = {
    'regressor__n_estimators': [100, 200],
    'regressor__max_depth': [None, 10, 20],
    'regressor__min_samples_split': [2, 5]
}

# GridSearchCV
grid = GridSearchCV(
    reg,
    param_grid,
    cv=5,
    scoring='neg_mean_squared_error'
)

grid.fit(X_train, y_train)

print("Best hyperparameters for Random Forest Regressor:")
print(grid.best_params_)
print(grid.best_score_)

Best hyperparameters for Random Forest Regressor:
{'regressor__max_depth': 10, 'regressor__min_samples_split': 2, 'regressor__n_estimators': 200}
-17.138409208761107


In [26]:
from sklearn.metrics import mean_squared_error, mean_absolute_error, r2_score
import numpy as np

best_model = grid.best_estimator_
y_test_pred = best_model.predict(X_test)

rmse = np.sqrt(mean_squared_error(y_test, y_test_pred))
mae = mean_absolute_error(y_test, y_test_pred)
r2 = r2_score(y_test, y_test_pred)

print("RMSE:", rmse)
print("MAE:", mae)
print("R2 Score:", r2)

RMSE: 4.001800582319817
MAE: 3.1050240801728437
R² Score: 0.921841374826045


# Evaluation Interpretation

## MAE = 3.105
* The average size of the errors were off by about 3.
* When the productivity is on the range of 35-98, a model with an error of about 3 is great.
* Large mistakes aren't penalized however.

## RMSE = 4.001
* Larger errors are penalized, thus large errors increase RMSE score.
* Because RMSE is higher than MAE, MAE failed to showcase bad and far off predictions.

## R2 = 0.9218
* The model does a good job at detecting the variances that determines worker productivity.
* 92% of the time, given the worker's features (e.g. education level, WFH hours, etc.) the model is able to accurately predict the productivity_score.

In [29]:
# Extract objects
preprocessor = best_model.named_steps['preprocessor']
regressor = best_model.named_steps['regressor']

# Original feature lists
num_features = preprocessor.transformers_[0][2]
cat_features = preprocessor.transformers_[1][2]

# Get transformed feature names
num_out = preprocessor.named_transformers_['num'].get_feature_names_out(num_features)
cat_out = preprocessor.named_transformers_['cat'].named_steps['onehot'].get_feature_names_out(cat_features)

# Combine to full feature list
feature_names = list(num_out) + list(cat_out)

# Importances
importances = regressor.feature_importances_

# Create DF
feature_importances_df = pd.DataFrame({
    'Feature': feature_names,
    'Importance': importances
}).sort_values(by='Importance', ascending=False)

feature_importances_df


Feature    Importance
7         Efficiency_Rating  5.084339e-01
5             Quality_Score  3.801523e-01
4      Task_Completion_Rate  6.034617e-02
6          Innovation_Score  1.114967e-02
9      Commute_Time_Minutes  2.563142e-03
..                      ...           ...
41       Job_Level_Director  4.242248e-06
125  Survey_Date_2024-03-04  3.996134e-06
94   Survey_Date_2024-02-02  3.207395e-06
80   Survey_Date_2024-01-19  1.995692e-06
126  Survey_Date_2024-03-05  9.261440e-07

[198 rows x 2 columns]